reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1319 @@
+	return 1
+}
+func (p *page) Start() bool {
+	return p.start
+}
+func (p *page) End() bool {
+	return p.end
+}
+func (p *page) Seq() Sequence {
+	return p.seq
+}
+func (p *page) IsPacket() bool {
+	return p.isPacket
+}
+func (p *page) String() string {
+	return fmt.Sprintf("page@%p{seq: %v, Bytes:%d, -> nextSeq:%v} (prev:%p, next:%p)", p, p.seq, len(p.bytes), p.seq+Sequence(len(p.bytes)), p.prev, p.next)
+}
+
+/* livePacket: implements a byteContainer */
+type livePacket struct {
+	bytes []byte
+	start bool
+	end   bool
+	ci    gopacket.CaptureInfo
+	ac    AssemblerContext
+	seq   Sequence
+}
+
+func (lp *livePacket) Bytes() []byte {
+	return lp.bytes
+}
+func (lp *livePacket) CaptureInfo() gopacket.CaptureInfo {
+	return lp.ci
+}
+func (lp *livePacket) AssemblerContext() AssemblerContext {
+	return lp.ac
+}
+func (lp *livePacket) Length() int {
+	return len(lp.bytes)
+}
+func (lp *livePacket) Start() bool {
+	return lp.start
+}
+func (lp *livePacket) End() bool {
+	return lp.end
+}
+func (lp *livePacket) Seq() Sequence {
+	return lp.seq
+}
+func (lp *livePacket) IsPacket() bool {
+	return true
+}
+
+// Creates a page (or set of pages) from a TCP packet: returns the first and last
+// page in its doubly-linked list of new pages.
+func (lp *livePacket) ConvertToPages(pc *pageCache, skip int, ac AssemblerContext) (*page, *page, int) {
+	ts := lp.ci.Timestamp
+	first := pc.next(ts)
+	current := first
+	current.prev = nil
+	first.isPacket = true
+	numPages := 1
+	seq, bytes := lp.seq.Add(skip), lp.bytes[skip:]
+	for {
+		length := min(len(bytes), pageBytes)
+		current.bytes = current.buf[:length]
+		copy(current.bytes, bytes)
+		current.seq = seq
+		current.ac = ac
+		bytes = bytes[length:]
+		if len(bytes) == 0 {
+			current.end = lp.End()
+			current.next = nil
+			break
+		}
+		seq = seq.Add(length)
+		current.next = pc.next(ts)
+		current.next.prev = current
+		current = current.next
+		numPages++
+	}
+	return first, current, numPages
+}
+func (lp *livePacket) estimateNumberOfPages() int {
+	return (len(lp.bytes) + pageBytes + 1) / pageBytes
+}
+
+func (lp *livePacket) Release(*pageCache) int {
+	return 0
+}
+
+// Stream is implemented by the caller to handle incoming reassembled
+// TCP data.  Callers create a StreamFactory, then StreamPool uses
+// it to create a new Stream for every TCP stream.
+//
+// assembly will, in order:
+//    1) Create the stream via StreamFactory.New
+//    2) Call ReassembledSG 0 or more times, passing in reassembled TCP data in order
+//    3) Call ReassemblyComplete one time, after which the stream is dereferenced by assembly.
+type Stream interface {
+	// Tell whether the TCP packet should be accepted, start could be modified to force a start even if no SYN have been seen
+	Accept(tcp *layers.TCP, ci gopacket.CaptureInfo, dir TCPFlowDirection, ackSeq Sequence, start *bool, ac AssemblerContext) bool
+	// ReassembledSG is called zero or more times.
+	// Reassembly objects are reused after each Reassembled call,
```

Comment:
Update comment to talk about ScatterGather/AssemblerContext instead of Reassembly obects.